### PR TITLE
Add CSRF-Support for FHEM

### PR DESCRIPTION
**Pull request by doublexminus**
**Adds the ability to set a CSRF-Token when configure an FHEM-Instance**

This CSRF-Token will be used to pull the list of Devices from the FHEM-Instance and it will be added to each FHEM-Devices which will be created from the FHEM-Devices-Tab by adding via "Build Item" or "Bulk Add".

Also fixes a problem with Quotations in generated URLs when using "Bulk Add" from FHEM-Devices-Tab.

### CODE REVIEW

**Comment by bwssytems** [comment]
a Question, is password used with FHEM in any case?

**Comment by doublexminus** [comment]
Since version 5.8 of FHEM the crsfToken is activated by default. But the Token will be updated after each reboot. Refers to [German FHEM crsfToken-HowTo](https://wiki.fhem.de/wiki/CsrfToken-HowTo)
If you set a fix Token by using:

> attr WEB.* csrfToken <any array of chars/digits>

Than you can use it with these changes.

**Comment by bwssytems** [comment]
I guess what I'm asking is if we use the password field as the crsfToken field instead of adding something else.

**Comment by doublexminus** [comment]
Oh, sorry... So you can also secure the web Interface with basicAuth, but it must be set explicitly.
I did not check the usage of password because I do not use it.

**Comment by bwssytems** [comment]
Is there a use for both at the same time or do you use basicAuth OR crsfToken?

**Comment by doublexminus** [comment]
I am going to check this on weekend but I guess it is possible to use both. basicAuth also secures the fhem-Webinterface while csrfToken ist only used for the API.

**Comment by bwssytems** [comment]
I'm just trying to make it simpler for the user if for the api crsfToken is the method that should be used, we shouldn't have both options available and just have the password field be the crsfToken.

**Comment by doublexminus** [comment]
So I checked it. 
If you have basicAuth activated for the FHEMWEB-"device" you have to set the user & password as well as the csrf-Token to the request. (e.g. `http://<user>:<pass>@<serverip>/fhem?fwcsrf=csrf_<token>&cmd=<fhemCommand>`)
It does not replace the authentication with username & password!

**Comment by bwssytems** [comment]
So I have mucked with the apps.js file in that latest release and now it is difficult to see your changes to that file. Could you rebase and update your branch and pull request please. Thanks

**Comment by bwssytems** [comment]
If there will be no rebase, I will close this pull request soon.

**Comment by jkellerer** [comment]
Hi @bwssytems, @doublexminus,

A CSRF token is indeed needed for quite a while in FHEM and disabling it is not recommended since it reduces security even more, so there is really a need to implement it to make the FHEM feature usable. However this PR is based on wrong assumptions and much too complex. It shouldn't be merged at all.

The CSRF token was misinterpreted as an access token or password, however CSRF tokens are there to protect from cross side request forgery (see [Wikipedia - CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery) ) . They are no passwords and they are never static so it makes no sense to enter them inside the GUI (unless you want to change it every 10 minutes or so).

Please see the shell script I'm using to call FHEM. It contains the logic required to call FHEM API methods:

```` .sh
#!/usr/bin/env bash
URL="https://USER:****@FHEMHOST/fhem"
XHR=$(curl -s -D - "$URL?XHR=1" | awk '/X-FHEM-csrfToken/{print $2}')
CMD="$*"
CMD=${CMD//+/%20}
CMD=${CMD// /%20}
echo "Calling FHEM $CMD"
curl --data "fwcsrf=$XHR" "$URL?cmd=$CMD"
````

Basically you get the CSRF token via a special API call. After that, you add the token as parameter for calls to any other API methods. If you don't get a CSRF token then likely CSRF tokens aren't required and can be skipped.

The call to "`https://myuser:mypw@myfhem/fhem?XHR=1`" returns the current token in the HTTP header, e.g. `X-FHEM-csrfToken: csrf_824173516473768`. 
After that you simply add a POST parameter named "`fwcsrf`" to all API calls (e.g. `fwcsrf=csrf_824173516473768`). Likely also works as GET parameter but I wouldn't use GET for API calls that are meant to change something.

Hope this helps.

**Comment by bwssytems** [comment]
Thanks for the input @jkellerer. Much appreciated.

**Comment by doublexminus** [comment]
Thanks @jkellerer for clearification!
